[ROM] Complement offline stage of ROM

Rbravo555 opened the pull request:
#depends on #9503

**📝 Description**
This PR adds the possibility to modify the parameters of a ROM simulation by editing the RomParameters.json.

In Particular, to indicate that a ROM simulation should be used for training the HROM elements and weights, it was necessary to pass a string to RomAnalysis. Now, one needs to set  "train_rom": True in the RomParameters.json.

**🆕 Changelog**
- Getting rid of the ElementsAndWeights.json file. Instead, its content is appended to the RomParameters.json
- Getting rid of the CreateHRomAnalysisInstance of the rom_analysis
- Modifying parameters of the rom_output_process to include train_hrom and run_hrom

Rbravo555 committed:
adding extra outputs to rom basis output process

Rbravo555 committed:
changing order to make it easier to manually change

Rbravo555 committed:
move functions of ECM to rom strategy. Append E and W to RomParameters

Rbravo555 committed:
Getting rid of ElementSelectionStrategy

Rbravo555 committed:
Expand RomParameters.json with training HROM flag

whether to train an HROM will be decided by reading the RomParameters.json, not by passing a string to the strategy

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
HRomTrainingUtility (missing model parts)

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Model parts creation

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Use ordered set with position insert

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Training working (missing visualization)

rubenzorrilla committed:
HROM seems to work (no visualization - no COMPUTE_HROM model part)

rubenzorrilla committed:
Skin visualization

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Clean up ROM B&S

rubenzorrilla committed:
Merge branch 'rom/clean-up-rom-bs' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
ROM always removes basis output process

rubenzorrilla committed:
Condition parents advances (not working yet)

rubenzorrilla committed:
ROM_SOLUTION_INCREMENT variable

rubenzorrilla committed:
Skin visualization and parents

rubenzorrilla committed:
Condition parents

rubenzorrilla committed:
Merge branch 'rom/hrom-in-rom-parameters' of https://github.com/KratosMultiphysics/Kratos into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Missing const in Has methods

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Advances

rubenzorrilla committed:
Independent HROM visualization mesh

rubenzorrilla committed:
Independent HROM visualization mesh

rubenzorrilla committed:
Merge branch 'rom/hrom-in-rom-parameters' of https://github.com/KratosMultiphysics/Kratos into rom/hrom-in-rom-parameters

SADPR commented on applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp:
```diff
@@ -0,0 +1,440 @@
+    for (auto it = r_elem_weights.begin(); it != r_elem_weights.end(); ++it) {
+        // Get element from origin mesh
+        const IndexType elem_id = stoi(it.name());
+        const auto p_elem = rOriginModelPart.pGetElement(elem_id + 1); //FIXME: WHY THIS +1?
```
Is it going to remain +1?  For me, it is ok for now...

rubenzorrilla replied:
I'd change the Ids 0 issue in a standalone PR to easily backtrack the changes if something gets messed up.

rubenzorrilla committed:
Merge branch 'master' into rom/hrom-in-rom-parameters

rubenzorrilla committed:
Minor fix

SADPR commented on applications/RomApplication/python_scripts/rom_analysis.py:
```diff
@@ -46,78 +86,67 @@ def ModifyAfterSolverInitialize(self):
+            if self.train_hrom:
+                self.__hrom_training_utility.AppendCurrentStepResiduals()
+
+            # #FIXME: Make this optional. This must be a process
```
I'm confused with this, isn't it always projected in ProjectToFineBasis?

rubenzorrilla replied:
The point is that the visualization model parts must be independent to the HROM one. I'll take care of this in an upcoming PR. So far I only included some tools required for the creation of such model parts.

SADPR reviewed (commented):
It looks good to me. @Rbravo555 check it so we can merge it.

rubenzorrilla committed:
Remove unused code

Rbravo555 committed:
Importing process to allow running ROM standalone

rubenzorrilla commented on applications/RomApplication/python_scripts/rom_analysis.py:
```diff
@@ -4,6 +4,7 @@
 import KratosMultiphysics.RomApplication as KratosROM
 from KratosMultiphysics.RomApplication import new_python_solvers_wrapper_rom
 from KratosMultiphysics.RomApplication.hrom_training_utility import HRomTrainingUtility
+from KratosMultiphysics.RomApplication.calculate_rom_basis_output_process import CalculateRomBasisOutputProcess
```
Why is this required? The process must be imported by the factory when instantiating the content of the `output_process_list`?

Rbravo555 replied:
I noticed the ROM simulation runs well right after the training. But when launching it separately (for changing the number of modes to use for example) Kratos complained that the “calculate_rom_output_process” didn’t belong to the RomApplication. Importing it manually here solved the problem for me..

rubenzorrilla replied:
OK. My bet is that if the process is not added in the process list it is not imported, then it cannot do the check below.

Rbravo555 committed:
Moving files to ROM and HROM folders . 

A flag must be switched in the RomParameters.json, this is the easy way of testing. To be improved in a future PR

Rbravo555 committed:
Getting rid of unused flags. Flags are now in the RomParameters.json

Rbravo555 committed:
Updating the test for the element selection algorithm.
